### ISSUE TITLE
A TypeScript recipe

### ISSUE THREAD

**Issue by goldbergyoni** [testjavascript/nodejs-integration-tests-best-practices#29]
Create a recipe with the same code as the example-ap, only with TypeScript, so both the SUT and the test code are typed.

Before picking this issue, it's recommended to have a short tech planning chat here and only then upon sync to get up to speed.

**Comment by DanielGluskin** [comment]
Does it mean we gonna maintain two apps in parallel or do we abolish the js version?

**Comment by goldbergyoni** [comment]
The production code can be kept as pure JS as our emphasis is on tests.

I would duplicate only the basic tests into a recipe and re-code with TS. Automating this might be a mess. I see value in serving the content to the reader in his favorite flavour: JS or TS.

@mikicho Suggests to modify all JS tests to TS so we have no JS in the basic tests. Less maintenance, not dramatic difference for the reader.

If we go with JS & TS, I would wait with the translation until the basic tests are done.

**Comment by mikicho** [comment]
@goldbergyoni just a minor correction: what I meant is to **duplicate** only the basic test to TS so we'll demonstrate only the relevant differences between TS and JS like `ts-node`, build,  and so on

**Comment by DanielGluskin** [comment]
been working on it a bit, should we use _ts-jest_?
also, I've written the basic-tests as .ts, doesn't look so different than .js, only the imports section changed.

**Comment by goldbergyoni** [comment]
Yes, ts-jest is a good default

Yes, should not be that different. Mostly the config file, imports, some typing, nothing too dramatic

@mikicho Yes, duplicating and not overriding the JS examples sounds good

**Comment by DanielGluskin** [comment]
Hey,

please look at [this](https://github.com/testjavascript/nodejs-integration-tests-best-practices/tree/typescript_recipeurl).

Added a new script called "typescript:test" and a new jest config file to use with it so it will compile only ts tests.

**Comment by goldbergyoni** [comment]
404 for me

**Comment by DanielGluskin** [comment]
Sorry, my mistake,
https://github.com/testjavascript/nodejs-integration-tests-best-practices/tree/typescript_recipe

**Comment by mikicho** [comment]
@goldbergyoni @DanielGluskin 
Now I'm thinking about it, why do we need to duplicate our `basic-tests` file?
js is a subset of ts, so we just need to run the tests with `tsjest` instead of `jest` to demonstrate the configurations and such.
WDYT?

**Comment by goldbergyoni** [comment]
I'm not sure.

As a TS developer, I'd love to see TS code (With types) in the tests, mocking is a bit different with TS. 

Not critical but nice to have.
